onchain/rollups/test/foundry/consensus/quorum/Quorum.t.sol

```diff
@@ -0,0 +1,293 @@
+
+    function testConstructorIgnoresDuplicates() external {
+        address[] memory validators = new address[](7);
+
+        validators[0] = vm.addr(1);
+        validators[1] = vm.addr(2);
+        validators[2] = vm.addr(1);
+        validators[3] = vm.addr(3);
+        validators[4] = vm.addr(2);
+        validators[5] = vm.addr(1);
+        validators[6] = vm.addr(3);
+
+        Quorum quorum = new Quorum(validators);
+
+        assertEq(quorum.numOfValidators(), 3);
+
+        for (uint256 i = 1; i <= 3; ++i) {
+            assertEq(quorum.validatorId(vm.addr(i)), i);
+            assertEq(quorum.validatorById(i), vm.addr(i));
+        }
+    }
+
+    function testValidatorId(uint8 numOfValidators, address addr) external {
+        address[] memory validators = generateAddresses(numOfValidators);
+
+        Quorum quorum = new Quorum(validators);
+
+        uint256 id = quorum.validatorId(addr);
+
+        if (contains(validators, addr)) {
+            assertLe(1, id);
+            assertLe(id, numOfValidators);
+        } else {
+            assertEq(id, 0);
+        }
+    }
+
+    function testValidatorById(uint8 numOfValidators, uint8 id) external {
+        Quorum quorum = _deployQuorum(numOfValidators);
+
+        address validator = quorum.validatorById(id);
+
+        if (id >= 1 && id <= numOfValidators) {
+            assertEq(quorum.validatorId(validator), id);
+        } else {
+            assertEq(validator, address(0));
+        }
+    }
+
+    function testSubmitClaimRevertsNotValidator(
+        uint8 numOfValidators,
+        address caller,
+        Claim calldata claim
+    ) external {
+        Quorum quorum = _deployQuorum(numOfValidators);
+
+        vm.assume(quorum.validatorId(caller) == 0);
+
+        vm.expectRevert("Quorum: caller is not validator");
+
+        vm.prank(caller);
+        quorum.submitClaim(claim);
+    }
+
+    function testNumOfValidatorsInFavorOf(
+        uint8 numOfValidators,
+        Claim calldata claim
+    ) external {
+        Quorum quorum = _deployQuorum(numOfValidators);
+        assertEq(quorum.numOfValidatorsInFavorOf(claim), 0);
+    }
+
+    function testIsValidatorInFavorOf(
+        uint8 numOfValidators,
+        Claim calldata claim,
+        uint256 id
+    ) external {
+        Quorum quorum = _deployQuorum(numOfValidators);
+        assertFalse(quorum.isValidatorInFavorOf(claim, id));
+    }
+
+    function testSubmitClaim(
+        uint8 numOfValidators,
+        Claim calldata claim
+    ) external {
+        numOfValidators = uint8(bound(numOfValidators, 1, 7));
+        Quorum quorum = _deployQuorum(numOfValidators);
+        bool[] memory submitted = new bool[](numOfValidators + 1);
+        for (uint256 id = 1; id <= numOfValidators; ++id) {
+            _submitClaimAs(quorum, claim, id);
+            submitted[id] = true;
+            _checkSubmitted(quorum, claim, submitted);
+        }
+    }
+
+    function testSubmitClaim300(Claim calldata claim) external {
```

Comment:
just out of curiosity, what this 300 means :)

Reply:
I will add a comment on this function to explain what it is testing exactly.